### PR TITLE
Code Quality: Convert PHP code base to use service container

### LINKED ISSUE #5292: Code Quality: Refactor Plugin class
There currently is a PHP in Plugin class has an ignore comment for the PHP MD as the class has too many properties. We should refactor the plugin class, as it doesn't use any many class properties.

### PR DESCRIPTION
**Pull request by spacedmonkey**
## Summary

Copy the infrastructure for service containers from amp plugin. Copy infrastructure and exception namespaces from amp plugin. Spin up classes in as services.  

## Relevant Technical Choices

Much of this PR is just changing the name of the init method to register and adding implements to all classes. 

The directories of Infrastructure and Exception are copied from the amp plugin. I had a couple of tweaks, including. 
- Adding `get_registration_action_priority` to `Delayed` interface to pick when a action is registered. 
- Lots of PHPStan fixes to phpdoc blocks. 
- Change namespace and added since phpdoc. 
- Add `REST_API_Factory` and `Register_Widget` classes, to move logic out of plugin class. 
- More methods in the plugin class ( to include files ) moved to namespace.php
- There an odd `is_needed` method in SVG class. Hopeful we remove this at some point. 
- Check is added to see if block is already registered, as in activate it is registered twice. 
- decoder is now injected in dashboard and story post type classes. 
- Improved Unit tests, so have helpers for adding / remove capabilities now that decoder is added to constructor of story post type. 
- Added some backwards compatibility with a magic `getter`. It is possible to get class properties like this, `get_plugin_instance()->customizer->render_stories()`. However it is no longer possible to do this, 

```
remove_action(  get_plugin_instance()->customizer, 'init );
```

For two reasons, 
1. The method init is renamed to register. 
2. This method is no longer register directory to action, it uses a closure. 

Plugin class still needs a little work. 

## To-do

- Add more tests.
- Update documentation. 

### QA

- [x] This is a non-user-facing change and requires no QA

This PR can be tested by following these steps:

1.

### UAT

- [x] UAT should use the same steps as above.

This PR can be tested by following these steps:

1.

## Checklist

- [ ] This PR addresses an existing issue and I have linked this PR to it in ZenHub
- [ ] I have tested this code to the best of my abilities
- [x] I have verified accessibility to the best of my abilities ([docs](https://github.com/google/web-stories-wp/blob/main/docs/accessibility-testiing.md))
- [x] I have verified i18n and l10n (translation, right-to-left layout) to the best of my abilities
- [x] This PR contains automated tests (unit, integration, and/or e2e) to verify the code works as intended ([docs](https://github.com/google/web-stories-wp/tree/main/docs#testing))
- [x] I have added documentation where necessary
- [x] I have added a matching `Type: XYZ` label to the PR

---

Fixes #5292

### CODE REVIEW

File: includes/Block/Web_Stories_Block.php [hunk 1]
```diff
@@ -343,4 +358,15 @@ protected function get_query_args() {
...
 
+	/**
+	 * Checks if a block type is registered.
+	 *
+	 * @since 1.6.0
+	 *
+	 * @param string $name Block type name including namespace.
+	 * @return bool True if the block type is registered, false otherwise.
+	 */
+	protected function is_register( $name ) {
```
Comment (swissspidy) [inline includes/Block/Web_Stories_Block.php]: is_registered\*

File: includes/Block/Web_Stories_Block.php [hunk 2]
```diff
@@ -84,6 +85,11 @@ public function init() {
 			$this->get_script_settings()
 		);
 
+		// Check to see if block is already registered.
+		if ( $this->is_register( self::BLOCK_NAME ) ) {
+			return;
+		}
+
```
Comment (swissspidy) [inline includes/Block/Web_Stories_Block.php]: Why is this needed?

File: tests/phpunit/tests/Story_Renderer/HTML.php [hunk 3]
```diff
@@ -125,7 +125,7 @@ public function test_replace_html_head_invalid() {
 	 */
 	public function test_add_publisher_logo() {
 		$attachment_id = self::factory()->attachment->create_upload_object( __DIR__ . '/../../data/attachment.jpg', 0 );
-		add_option( Settings::SETTING_NAME_ACTIVE_PUBLISHER_LOGO, $attachment_id );
+		update_option( Settings::SETTING_NAME_ACTIVE_PUBLISHER_LOGO, $attachment_id );
```
Comment (swissspidy) [inline tests/phpunit/tests/Story_Renderer/HTML.php]: Why was this change needed? Sounds like we're not properly cleaning that up somewhere else.

File: includes/namespace.php [hunk 4]
```diff
@@ -129,10 +130,8 @@ function remove_site( $error, $site ) {
  * @return void
  */
 function deactivate( $network_wide ) {
-	unregister_post_type( Story_Post_Type::POST_TYPE_SLUG );
-	if ( ! defined( '\WPCOM_IS_VIP_ENV' ) || false === \WPCOM_IS_VIP_ENV ) {
-		flush_rewrite_rules( false ); // phpcs:ignore WordPressVIPMinimum.Functions.RestrictedFunctions.flush_rewrite_rules_flush_rewrite_rules
-	}
```
Comment (swissspidy) [inline includes/namespace.php:135]: This code was removed without replacement.

It ensured that rewrite rules are properly flushed upon plugin deactivation.

Related: why aren't we doing that on plugin _activation_ as well? 🤔

Comment (spacedmonkey) [reply includes/namespace.php:135]: This happens here now. 

https://github.com/google/web-stories-wp/pull/6906/files#diff-afc0298087d72ee3c18032663a02b2c9f484a7bd1262fe8956c0300b8fb0305cR141-R143

It does now do that on deactivation. 

https://github.com/google/web-stories-wp/pull/6906/files#diff-afc0298087d72ee3c18032663a02b2c9f484a7bd1262fe8956c0300b8fb0305cR163-R165

File: tests/phpunit/tests/Story_Renderer/HTML.php (hunk 3 above)
Comment (spacedmonkey) [reply tests/phpunit/tests/Story_Renderer/HTML.php]: I don't know why this change is needed, it seems unrelated. I will dig into why.

File: includes/Block/Web_Stories_Block.php (hunk 2 above)
Comment (spacedmonkey) [reply includes/Block/Web_Stories_Block.php]: > Check is added to see if block is already registered, as in activate it is registered twice.

It because the services are register again on activation.

**Comment by spacedmonkey** [comment]
CC @schlessera and @westonruter 

I have whole sale copied your implementation of service containers in your plugin to bring uniformity to the two plugins. The implementation is solid and generic. Baring a couple of tweaks and change of namespaces, went in super smooth.

Love your thoughts and feedback on this PR.

File: includes/namespace.php (hunk 4 above)
Comment (swissspidy) [reply includes/namespace.php:135]: Sorry, these links don't seem to work for me. But I found it now in `ServiceBasedPlugin`

File: includes/Block/Web_Stories_Block.php (hunk 1 above)
Comment (spacedmonkey) [reply includes/Block/Web_Stories_Block.php]: Fixed

**Comment by swissspidy** [comment]
SVG upload E2E test is constantly failing. Does it work for you? Might be a bug.

**Comment by schlessera** [comment]
@spacedmonkey I'll see if I can get a more detailed review in soon, but I already wanted to give a few pointers about the general approach.

#### Timing

I see you've added a required action of `init` and a priority of 10 to everything. Regarding the priority, 10 is the default, so this is redundant. Regarding the default action, it doesn't make sense to me to specify that everywhere via the `Delayed` interface, as this is basically a "double-hook" process. You're calling the `register()` method on the `init` action, only to then decide what hooks to actually register the service functionality on. So, ideally, you'd just `register()` immediately by default and have the `register()` method decide by its hooks what the timing is supposed to be.

So, instead of:
```
instantiate() => {
	'init' => register() {
		'some_action' => $this->callback()
	}
}
```
it would be:
```
instantiate() => register() {
	'some_action' => $this->callback()
}
```

`Delayed` is meant for scenarios where the instantiation of the service can only happen at a certain point in time.
If all of your services require the `'init'` action to be able to even instantiate, then you change the registration action here: https://github.com/google/web-stories-wp/pull/6906/files#diff-afc0298087d72ee3c18032663a02b2c9f484a7bd1262fe8956c0300b8fb0305cR49

#### Service Base class

I would recommend against using a shared base class for services. If you need to do something that is true for all services, it can probably be re-thought and done in some other way, like with the above timing change. And if you still want to use a base class for some shared functionality, I'd recommend naming it for what it actually provides, instead of "replacing" the interface, which can cause confusion.

#### Make use of the Injector

The infrastructure is mostly built around the principle that the instantiation should happen in a central, controlled way, where the ideal is to have one instantiation call instantiate the entire object graph as needed.

This can be achieved by making use of dependency injection, which has a recursive effect.

For the rest of the code, the `new` keyword should serve as a code smell.

Whenever you find the need to use the `new` keyword, you should:
1. Check whether you can have that object just injected via the constructor;
2. Otherwise, get a hold of the `Injector` to instantiate it (while having all dependencies be dealt with automatically).

For 2., this would look like this:
```
$story_post_type = new Story_Post_Type( new Experiments(), new Meta_Boxes(), new Decoder() );
```
you should have something like this:
```
$story_post_type = $injector->make( Story_Post_Type::class );
```
Note that you don't need to know anything about the dependencies, they are controlled via the injector.

Getting a hold of the injector is done via:
1. constructor injection :)
2. `$container->get( 'injector' )`
3. `Services::get_injector()`

Hope that helps, and I'm happy to answer any questions.

**Comment by schlessera** [comment]
Hmm, looking at it a bit more closely, it seems like you're not yet using the service-based architecture...?

The core of the infrastructure is the `ServiceBasedPlugin` abstract class, which would need to be implemented by something like a `WebStoriesPlugin` class, and that would be what you trigger in your bootstrap file. Is this still WIP and I'm just too early with the feedback?

**Comment by schlessera** [comment]
You can see the core of the AMP plugin here, for example: https://github.com/ampproject/amp-wp/blob/develop/src/AmpWpPlugin.php

**Comment by spacedmonkey** [comment]
> $story_post_type = new Story_Post_Type( new Experiments(), new Meta_Boxes(), new Decoder() );

There are only a handful of places where this is used now, I can refactor them to use the injector. 

>I see you've added a required action of init and a priority of 10 to everything. 

This is not true, there are a number of classes, that have different actions and properties. It defaults to these in the service, but are overridden where required. 

> Delayed is meant for scenarios where the instantiation of the service can only happen at a certain point in time.
If all of your services require the 'init' action to be able to even instantiate, then you change the registration action here:

All of current classes are written with this delay in mind. So to keep to the current format and not to introduce to many changes, I kept it this way. I think should maybe look into the changing this later, as this is already a big refactor. 

>  it doesn't make sense to me to specify that everywhere via the Delayed interface, as this is basically a "double-hook" process

Why? Can you explain this a little more detail. I don't see it doubling hooking. Is it double hooking here?

```
foreach ( $services as $id => $class ) {
			$id    = $this->maybe_resolve( $id );
			$class = $this->maybe_resolve( $class );
			// Allow the services to delay their registration.
			if ( is_a( $class, Delayed::class, true ) ) {
				$registration_action = $class::get_registration_action();
				if ( did_action( $registration_action ) ) {
					$this->register_service( $id, $class );

					continue;
				}

				\add_action(
					$class::get_registration_action(),
					function () use ( $id, $class ) {
						$this->register_service( $id, $class );
					},
					$class::get_registration_action_priority()
				);

				continue;
			}

			$this->register_service( $id, $class );
		}
```

> You can see the core of the AMP plugin here, for example: https://github.com/ampproject/amp-wp/blob/develop/src/AmpWpPlugin.php

Plugin has been changed to extend - `ServiceBasedPlugin`. Without that the code wouldn't work at all.

**Comment by spacedmonkey** [comment]
> SVG upload E2E test is constantly failing. Does it work for you? Might be a bug.

I am seeing the error locally, but I don't see how it would be related to my change...

**Comment by schlessera** [comment]
> Why? Can you explain this a little more detail. I don't see it doubling hooking.

Yes, sure. So, consider a Service with the `register()` method and a `process()` method that needs to run on the `'wp'` action. The `register()` method of that service will contain the `add_action()` call to hook the method of the service to the WordPress lifecycle: `add_action( 'wp', [ $this, 'process' ] );`.

If you now use a `Delayed` interface to that service and tell it to use the `init` action as the registration hook, then you are actually doing an `add_action()` to even call the `register()` method in the first place.

So instead of this:
1. instantiate service
2. execute `register()`
3. `add_action( 'wp', [ $this, 'process' ] );`
4. forward to `'wp'` hook
5. execute `[ $this, 'process' ]`

You will end up with this:
1. instantiate service
2. `add_action( 'init', [ $this, 'register' ] );`
3. forward to `'init'` action
4. execute `register()`
5. `add_action( 'wp', [ $this, 'process' ] );`
6. forward to `'wp'` hook
7. execute `[ $this, 'process' ]`

Note that you have two hooks (2. & 5.) just to execute `[ $this, 'process' ]` at the right moment.

**Comment by spacedmonkey** [comment]
You are talking about the hooking the register method on init and then within that method, hooking into add other methods? 

If so, I know this is not the best, but this is how this currently works in our plugin. For example 

https://github.com/google/web-stories-wp/blob/05ba6811cae5633354358551fb67e91838d837b9/includes/Plugin.php#L323
https://github.com/google/web-stories-wp/blob/05ba6811cae5633354358551fb67e91838d837b9/includes/Dashboard.php#L100-L106

I know this isn't great, but there are reasons we have done this in our code base. 

I really don't want to tackle this problem in this PR. I will change this in a follow up PR.

File: includes/REST_API_Factory.php [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+		$templates_lock->register_routes();
+
+		$stories_lock = new Stories_Lock_Controller( Story_Post_Type::POST_TYPE_SLUG );
+		$stories_lock->register_routes();
+
+		$stories_media = new Stories_Media_Controller( 'attachment' );
+		$stories_media->register_routes();
+
+		$stories_users = new Stories_Users_Controller();
+		$stories_users->register_routes();
```
Comment (schlessera) [inline includes/REST_API_Factory.php:77]: These could also be turned into a series of service. The `register_routes()` method could just as well be a simple `register()`.

The constructor arguments for some of these can be set via (overriding) `ServiceBasedPlugin::get_arguments()`.

File: includes/SVG.php [hunk 6]
```diff
@@ -117,6 +92,16 @@ public function init() {
...
 	}
 
+	/**
+	 * Check whether the conditional object is currently needed.
+	 *
+	 * @return bool Whether the conditional object is needed.
+	 */
+	public static function is_needed() {
+		$experiments = new Experiments();
+		return $experiments->is_experiment_enabled( 'enableSVG' );
```
Comment (schlessera) [inline includes/SVG.php]: I'd recommend using constants for the experiments names, otherwise a simple typo will make something like the above behave erratically:
```suggestion
		return $experiments->is_experiment_enabled( Experiment::ENABLE_SVG );
```

File: includes/Story_Post_Type.php [hunk 7]
```diff
@@ -217,6 +216,27 @@ public function init() {
...
+
+	/**
+	 * Deactivate the service.
+	 *
+	 * @param bool $network_wide Whether the deactivation was done network-wide.
+	 * @return void
+	 */
+	public function deactivate( $network_wide ) {
+		$this->remove_caps_from_roles();
+		unregister_post_type( self::POST_TYPE_SLUG );
```
Comment (schlessera) [inline includes/Story_Post_Type.php:246]: This is probably not needed, and could lead to problems in other parts of the deactivation routine, if they somehow interact with the post type for clean-up or so.
```suggestion
```

File: includes/functions.php [hunk 8]
```diff
@@ -63,6 +63,7 @@ function get_stories( array $attrs = [], array $query_args = [] ) {
  * @return void
  */
 function render_theme_stories() {
+	$customizer = new Customizer();
 	//phpcs:disable WordPress.Security.EscapeOutput.OutputNotEscaped
-	echo get_plugin_instance()->customizer->render_stories();
+	echo $customizer->render_stories();
```
Comment (schlessera) [inline includes/functions.php]: This could potentially be rewritten like this:
```php
echo Services::get( 'customizer' )->render_stories();
```

This avoids relying on `new` and thus keeps all indirection and dependency management intact.

Comment (schlessera) [reply includes/functions.php]: Note in case the static access to `Services::get()` looks problematic:
This is not ideal, but still better than `new Customizer()`. Using `new` is the same type of static access, only to `Customizer::__construct()` instead. However, what makes the solution with `new` worse is that you not only hard-coded a static access, but you also hard-coded the arguments for the constructor (so, how its dependencies are handled), and you hard-coded the way the instantiation is optimized and/or shared (i.e. not at all).

File: includes/namespace.php [hunk 9]
```diff
@@ -38,17 +38,18 @@
  * @return void
  */
 function setup_new_site() {
-	$story = new Story_Post_Type( new Experiments(), new Meta_Boxes() );
-	$story->init();
+	$story = new Story_Post_Type( new Experiments(), new Meta_Boxes(), new Decoder() );
+	$story->register();
```
Comment (schlessera) [inline includes/namespace.php]: Could be (for a later PR, as discussed):
```php
Services::get( 'story_post_type' )->register()
```

File: includes/namespace.php [hunk 10]
```diff
@@ -38,17 +38,18 @@
...
 	$story->add_caps_to_roles();
 	if ( ! defined( '\WPCOM_IS_VIP_ENV' ) || false === \WPCOM_IS_VIP_ENV ) {
 		flush_rewrite_rules( false ); // phpcs:ignore WordPressVIPMinimum.Functions.RestrictedFunctions.flush_rewrite_rules_flush_rewrite_rules
 	}
 
 	$database_upgrader = new Database_Upgrader();
-	$database_upgrader->init();
+	$database_upgrader->register();
```
Comment (schlessera) [inline includes/namespace.php]: Could be (for a later PR, as discussed):
```php
Services::get( 'database_upgrader' )->register()
```

File: includes/namespace.php [hunk 11]
```diff
@@ -112,7 +113,7 @@ function remove_site( $error, $site ) {
 		return;
 	}
 	$site_id = (int) $site->blog_id;
-	$story   = new Story_Post_Type( new Experiments(), new Meta_Boxes() );
+	$story   = new Story_Post_Type( new Experiments(), new Meta_Boxes(), new Decoder() );
 	switch_to_blog( $site_id );
 	$story->remove_caps_from_roles();
```
Comment (schlessera) [inline includes/namespace.php]: Similar:
```suggestion
	switch_to_blog( $site_id );
	Services::get( 'story_post_type' )->remove_caps_from_roles();
```

**Review by schlessera — DISMISSED** [review]
The naming scheme for classes is now mixed between `WordPress_Underscore_Class_Naming` and `RegularPsr4ClassNaming` and should maybe cleaned-up (purely optional, though).

**Comment by schlessera** [comment]
Also, more generally, there's a big Google copyright plastered across all of the infrastructure code and tests, but not a single attribution to the real source of this: https://github.com/mwpd/basic-scaffold

A short mention for source attribution somewhere (if only in the readme) would be appreciated.

File: includes/SVG.php (hunk 6 above)
Comment (spacedmonkey) [reply includes/SVG.php]: Not really related to the change in hand. I also disagree, as there are lots and lots of feature flags and that would become messy.

File: includes/REST_API_Factory.php (hunk 5 above)
Comment (spacedmonkey) [reply includes/REST_API_Factory.php:77]: Yeah, I thought this, but for now, I think I am going to leave it as it is and add a todo to the code.

File: includes/Story_Post_Type.php (hunk 7 above)
Comment (spacedmonkey) [reply includes/Story_Post_Type.php:246]: This is how it worked before and I am not going to make functional changes, just refactor the code. It must have been needed at some, which is why that is there. I am not going to second guess this for now.

File: includes/functions.php (hunk 8 above)
Comment (spacedmonkey) [reply includes/functions.php]: As this is a function that is not even used in our codebase,I thought a new class is not really a problem here. I will convert to use the injector.

**Comment by spacedmonkey** [comment]
@schlessera Please can you refrain from comments out of scope of this change. This is not a full audit of our code. I am just trying to shoehorn in service containers to the codebase. I am not currently second guessing how the classes are written, just trying to get them to work as is. Later we can / will refactor to improve, that as this PR is already large, I do not wish to make this part of this PR. 

Any references to Database_Upgrader and Story_Post_Type have been avoid. There are tickets to refactor these classes https://github.com/google/web-stories-wp/issues/5062 https://github.com/google/web-stories-wp/issues/5668. We will tackle changes to how those classes are spun and called in those tickets.

File: includes/namespace.php (hunk 11 above)
Comment (spacedmonkey) [reply includes/namespace.php]: Let's handle this here https://github.com/google/web-stories-wp/issues/5062

File: includes/namespace.php (hunk 10 above)
Comment (spacedmonkey) [reply includes/namespace.php]: Let's handle this https://github.com/google/web-stories-wp/issues/5668

File: includes/namespace.php (hunk 9 above)
Comment (spacedmonkey) [reply includes/namespace.php]: Let's handle this here https://github.com/google/web-stories-wp/issues/5062

File: includes/functions.php (hunk 8 above)
Comment (spacedmonkey) [reply includes/functions.php]: fixed.

File: includes/functions.php [hunk 12]
```diff
@@ -63,7 +63,12 @@ function get_stories( array $attrs = [], array $query_args = [] ) {
  * @return void
  */
 function render_theme_stories() {
-	$customizer = new Customizer();
+	$injector = Services::get_injector();
+	if ( ! method_exists( $injector, 'make' ) ) {
+		return;
+	}
```
Comment (schlessera) [inline includes/functions.php:69]: Are these `method_exists()` checks used for static analysis?

If so, you can actually adapt static analysis to work transparently. Here's how I did that for PHPStan:
https://github.com/ampproject/amp-wp/tree/develop/tests/php/src/PhpStan

Also, you can enable auto-completion and proper analysis in PHPStorm via a PHPStorm meta file:
https://github.com/ampproject/amp-wp/blob/develop/.phpstorm.meta.php

Comment (schlessera) [reply includes/functions.php:69]: Also, keep in mind that performance-wise, it is preferable to use the container instead of the injector in cases like this, as the container is a simple hash lookup.

Comment (spacedmonkey) [reply includes/functions.php:69]: Wow, this is cool. Great work. I think for now, let's just get this merged and I will follow up with copying this change as well.

File: includes/SVG.php (hunk 6 above)
Comment (swissspidy) [reply includes/SVG.php]: @spacedmonkey Could you create a ticket for this suggestion? :-)

File: includes/Story_Post_Type.php (hunk 7 above)
Comment (swissspidy) [reply includes/Story_Post_Type.php:246]: IIRC we need to unregister the post type before we can flush the permalinks. Otherwise the rewrites for that post type won't be removed.

**Comment by swissspidy** [comment]
> Also, more generally, there's a big Google copyright plastered across all of the infrastructure code and tests, but not a single attribution to the real source of this: [mwpd/basic-scaffold](https://github.com/mwpd/basic-scaffold)
> 
> A short mention for source attribution somewhere (if only in the readme) would be appreciated.

Absolutely 👍  While we don't have an attribution section in the top-level readme _yet_, we can  add appropriate notes in the main Service + Plugin classes as a starter.

**Comment by spacedmonkey** [comment]
@schlessera You are right, copyright was a bad call here. I have removed it and added some props. https://github.com/google/web-stories-wp/pull/6906/commits/8d15d187314888e11b1daf3a9b8979f8ae4687b8

If anyone has any ideas on better wording or format, it is now a single find and replace.

**Comment by swissspidy** [comment]
> @schlessera You are right, copyright was a bad call here. I have removed it and added some props. [8d15d18](https://github.com/google/web-stories-wp/commit/8d15d187314888e11b1daf3a9b8979f8ae4687b8)
> 
> If anyone has any ideas on better wording or format, it is now a single find and replace.

We have some guidelines:

Files that copy lines of code from third-party files, or that are directly translated or derived from third-party files, must retain any copyright notices present in the third-party file(s). A Google copyright notice can be added to the license header as a separate copyright notice, beneath any existing notices, but above the license text.

Any files that are exact copies of files taken from a third-party source must retain any existing copyright notices and must not include any new copyright notices.

**Comment by spacedmonkey** [comment]
Should we be copying [this](https://github.com/mwpd/basic-scaffold/blob/4e2d4cff221056ed131226b9cc4695ddca36361f/src/Infrastructure/Deactivateable.php#L3-L11) then?

**Comment by swissspidy** [comment]
For new files that are not found in `basic-scaffold` we obviously don't need to add something custom and can use the regular copyright notice.

For files originating from `basic-scaffold` _without modification_ we can use:

```php
/**
 * File description.
 *
 * @package   Google\Web_Stories
 * @author    Alain Schlesser <alain.schlesser@gmail.com>
 * @license   MIT
 * @link      https://www.mwpd.io/
 * @copyright 2019 Alain Schlesser
 */
```

For files originating from `basic-scaffold` but _modified_ we can use:

```php
/**
 * File description.
 *
 * @package   Google\Web_Stories
 * @copyright 2019 Alain Schlesser
 * @license   MIT
 * @link      https://www.mwpd.io/
 */

/**
 * Original code modified for this project.
 *
 * @copyright 2020 Google LLC
 * @license   https://www.apache.org/licenses/LICENSE-2.0 Apache License 2.0
 */
```

**Comment by spacedmonkey** [comment]
Updated comments. https://github.com/google/web-stories-wp/pull/6906/commits/14c2f49a64b74cc4d7ab59a9b3d98b09d74d0fdc

File: includes/Exception/FailedToMakeInstance.php [hunk 13]
```diff
@@ -0,0 +1,177 @@
...
+namespace Google\Web_Stories\Exception;
+
+use Google\Web_Stories\Infrastructure\Injector\InjectionChain;
+use RuntimeException;
+
+/**
+ * Exception thrown when the injector couldn't instantiate a given class or
+ * interface.
+ *
+ * @since 1.6
```
Comment (swissspidy) [inline includes/Exception/FailedToMakeInstance.php]:
```suggestion
 * @since 1.6.0
```

File: includes/Exception/FailedToMakeInstance.php [hunk 14]
```diff
@@ -0,0 +1,177 @@
+<?php
+/**
+ * Exception FailedToMakeInstance.
+ *
+ * @package Google\Web_Stories
+ */
```
Comment (swissspidy) [inline includes/Exception/FailedToMakeInstance.php]: The regular file description docblocks in the plugiin look more like this

```php
/**
 * Exception FailedToMakeInstance.
 *
 * @package   Google\Web_Stories
 * @copyright 2020 Google LLC
 * @license   https://www.apache.org/licenses/LICENSE-2.0 Apache License 2.0
 * @link      https://github.com/google/web-stories-wp
 */
```

File: includes/Embed_Base.php [hunk 15]
```diff
@@ -83,6 +83,15 @@ public function init() {
 		add_filter( 'wp_kses_allowed_html', [ $this, 'filter_kses_allowed_html' ], 10, 2 );
 	}
 
+	/**
+	 * Get the action priority to use for registering the service.
+	 *
```
Comment (swissspidy) [inline includes/Embed_Base.php:88]: New methods like this would need `@since` annotations

File: includes/Infrastructure/Activateable.php [hunk 16]
```diff
@@ -0,0 +1,43 @@
...
+/**
+ * Something that can be activated.
+ *
+ * By tagging a service with this interface, the system will automatically hook
+ * it up to the WordPress activation hook.
+ *
+ * This way, we can just add the simple interface marker and not worry about how
+ * to wire up the code to reach that part during the static activation hook.
+ *
+ * @since 1.6
```
Comment (swissspidy) [inline includes/Infrastructure/Activateable.php]:
```suggestion
 * @since 1.6.0
```

let's use this full version number everywhere for consistency, so `1.6.0` instead of just `1.6`.

File: includes/Exception/WebStoriesException.php [hunk 17]
```diff
@@ -0,0 +1,38 @@
+<?php
+/**
+ * Interface WebStoriesException.
+ *
+ * @package Google\Web_Stories
```
Comment (swissspidy) [inline includes/Exception/WebStoriesException.php]: Ditto about the file docblock here

File: includes/Register_Widget.php [hunk 18]
```diff
@@ -0,0 +1,52 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+namespace Google\Web_Stories;
+
+/**
+ * Class RegisterWidget
```
Comment (swissspidy) [inline includes/Register_Widget.php:30]:
```suggestion
 * Class Register_Widget
```

**Review by swissspidy — APPROVED** [review]
Only a few nits here and there, otherwise LGTM 👍 

Needs follow-up tickets/PRs to address remaining feedback from code review.

**Comment by spacedmonkey** [comment]
Follow tasks. 

- Use injector / contain instead of creating new instances 
- Implement PHPStorm fix. 
- Implement PHPStan look via injector
- Use injector in REST API factory

File: includes/Register_Widget.php (hunk 18 above)
Comment (spacedmonkey) [reply includes/Register_Widget.php:30]: Fixed

File: includes/Exception/WebStoriesException.php (hunk 17 above)
Comment (spacedmonkey) [reply includes/Exception/WebStoriesException.php]: Fixed

File: includes/Infrastructure/Activateable.php (hunk 16 above)
Comment (spacedmonkey) [reply includes/Infrastructure/Activateable.php]: Fixed

File: includes/Embed_Base.php (hunk 15 above)
Comment (spacedmonkey) [reply includes/Embed_Base.php:88]: Fixed

File: includes/Exception/FailedToMakeInstance.php (hunk 14 above)
Comment (spacedmonkey) [reply includes/Exception/FailedToMakeInstance.php]: Fixed

File: includes/Exception/FailedToMakeInstance.php (hunk 13 above)
Comment (spacedmonkey) [reply includes/Exception/FailedToMakeInstance.php]: Fixed

**Comment by swissspidy** [comment]
One merge conflict here now before we can 🚢  it

**Review by swissspidy — APPROVED** [review]
